heron/metricscachemgr/tests/java/com/twitter/heron/metricscachemgr/metricscache/CacheCoreTest.java

```diff
@@ -0,0 +1,741 @@
+    // there is 0 <component, instance, metric> tuples
+    assertEquals(response.getMetricList().size(), 0);
+  }
+
+  private void prepareDataForTreeIndex() {
+    // create cache with time window 100 seconds, bucket size 30 seconds and no exception store.
+    // the cache should be initialized with 4 buckets:
+    // bucket 1: [now-100 seconds ~ now-70 seconds)
+    // bucket 2: [now-70 seconds ~ now-40 seconds)
+    // bucket 3: [now-40 seconds ~ now-10 seconds)
+    // bucket 4: [now-10 seconds ~ now]
+    cacheCore = new CacheCore(100, 30, 0);
+
+    // current timestamp used as time origin
+    // although it may be slightly different from the time origin
+    // in the CacheCore initialization.
+    now = System.currentTimeMillis();
+    TopologyMaster.PublishMetrics.Builder builder = TopologyMaster.PublishMetrics.newBuilder();
+    long[] ts = new long[]{
+        // the timestamp falls outside cache time window. too old to be in the cache
+        now - 120 * 1000,
+        // should be in bucket 1
+        now - 90 * 1000,
+        // should be in bucket 1
+        now - 80 * 1000,
+        // should be in bucket 2
+        now - 60 * 1000,
+        // should be in bucket 2
+        now - 50 * 1000,
+        // should be in bucket 3
+        now - 30 * 1000,
+        // should be in bucket 3
+        now - 20 * 1000,
+        // should be in bucket 4
+        now
+    };
+    String[] vals = new String[]{
+        "0.0", "0.1", "0.2", "0.3", "0.4", "0.5", "0.6", "0.7"
+    };
+    for (int i = 0; i < 8; i++) {
+      builder.addMetrics(TopologyMaster.MetricDatum.newBuilder()
+          .setTimestamp(ts[i])
+          .setComponentName("c1").setInstanceId("i1")
+          .setName("m1")
+          .setValue(vals[i]));
+    }
+
+    cacheCore.addMetricException(builder.build());
+
+    // initialization
+    metricsFilter = new MetricsFilter();
+    metricsFilter.setMetricToType("m1", MetricsFilter.MetricAggregationType.SUM);
+
+    timeRangeValueComparator =
+        new Comparator<MetricTimeRangeValue>() {
+          @Override
+          public int compare(MetricTimeRangeValue o1,
+                             MetricTimeRangeValue o2) {
+            return (int) (o1.getStartTime() - o2.getStartTime());
+          }
+        };
+  }
+
+  /*
+   * query 1 bucket
+   */
+  @Test
+  public void testTreeIndex1() {
+    prepareDataForTreeIndex();
+
+    long startTime = now - 95 * 1000;
+    long endTime = now - 75 * 1000;
+    MetricRequest request = new MetricRequest(null, null, startTime, endTime, RAW);
+    MetricResponse response = cacheCore.getMetrics(request, metricsFilter);
+    // there is only one <component, instance, metric> tuple
+    List<MetricDatum> metricList = response.getMetricList();
+    assertEquals(metricList.size(), 1);
+    // there should be 2 metrics
+    List<MetricTimeRangeValue> list =
+        metricList.get(0).getMetricValue();
+    assertEquals(list.size(), 2);
```

Comment:
instead of hardcoding 2, can you assert size in the assertMetricValue method based on expectedList.size()? Same for all other tests in this class.